src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -0,0 +1,142 @@
+                _coreTypes.Add(name, type);
+            }
+
+            return type;
+        }
+
+        internal void AppendMetadata(MetadataBuilder metadata)
+        {
+            // Add module metadata
+            metadata.AddModule(
+                generation: 0,
+                metadata.GetOrAddString(ScopeName),
+                metadata.GetOrAddGuid(Guid.NewGuid()),
+                default,
+                default);
+
+            // Create type definition for the special <Module> type that holds global functions
+            metadata.AddTypeDefinition(
+                default,
+                default,
+                metadata.GetOrAddString("<Module>"),
+                baseType: default,
+                fieldList: MetadataTokens.FieldDefinitionHandle(1),
+                methodList: MetadataTokens.MethodDefinitionHandle(1));
+
+            // Add each type definition to metadata table.
+            foreach (TypeBuilderImpl typeBuilder in _typeDefStore)
+            {
+                TypeReferenceHandle parent = default;
+                if (typeBuilder.BaseType is not null)
+                {
+                    // TODO: need to handle the case when the base is from same assembly
+                    parent = GetTypeReference(metadata, typeBuilder.BaseType);
+                }
+
+                TypeDefinitionHandle typeDefinitionHandle = MetadataHelper.AddTypeDefinition(metadata, typeBuilder, parent, _nextMethodDefRowId, _nextFieldDefRowId);
+
+                // Add each method definition to metadata table.
+                foreach (MethodBuilderImpl method in typeBuilder._methodDefStore)
+                {
+                    MetadataHelper.AddMethodDefinition(metadata, method, method.GetMethodSignatureBlob());
+                    _nextMethodDefRowId++;
+                }
+
+                foreach (FieldBuilderImpl field in typeBuilder._fieldDefStore)
+                {
+                    MetadataHelper.AddFieldDefinition(metadata, field);
+                    _nextFieldDefRowId++;
+                }
+            }
+        }
+
+        private TypeReferenceHandle GetTypeReference(MetadataBuilder metadata, Type type)
+        {
+            if (!_typeRefStore.TryGetValue(type, out var parentHandle))
+            {
+                parentHandle = MetadataHelper.AddTypeReference(metadata, type,
+                    GetAssemblyReference(type.Assembly, metadata));
+            }
+
+            return parentHandle;
+        }
+
+        private AssemblyReferenceHandle GetAssemblyReference(Assembly assembly, MetadataBuilder metadata)
+        {
+            if (_assemblyRefStore.TryGetValue(assembly, out var handle))
+            {
+                return handle;
+            }
+
+            return MetadataHelper.AddAssemblyReference(assembly, metadata);
+        }
+        [RequiresAssemblyFiles("Returns <Unknown> for modules with no file path")]
+        public override string Name => _name;
```

Comment:
```suggestion
        public override string Name => "<In Memory Module>";
```
This is what .NET Framework returns as a name here.